File: src/main/java/gargoyle/ct/task/impl/CTTimer.java
package gargoyle.ct.task.impl;

import gargoyle.ct.config.CTConfig;
import gargoyle.ct.task.CTTaskUpdatable;
import gargoyle.ct.task.helper.CTTimeHelper;

import java.util.Arrays;
import java.util.Timer;

public class CTTimer {
    private static final int CHECK_DELAY = 100;
    private static final int CHECK_PERIOD = 500;
    private final Timer timer;
    private final CTTimerTask timerTask;

    public CTTimer(CTTimeHelper timeHelper, CTTaskUpdatable... updatables) {
        this(timeHelper, Arrays.asList(updatables));
    }

    public CTTimer(CTTimeHelper timeHelper, Iterable<CTTaskUpdatable> updatables) {
        timer = new Timer(CTTimer.class.getName(), true);
        timerTask = new CTTimerTask(timeHelper, updatables);
        timer.scheduleAtFixedRate(timerTask, CHECK_DELAY, CHECK_PERIOD);
    }

    public void arm(CTConfig config, long currentMillis) {
        CTTask task = timerTask.getTask();
        task.setConfig(config);
        task.setStarted(currentMillis);
    }

    public void terminate() {
        timer.cancel();
    }

    public void unarm() {
        CTTask task = timerTask.getTask();
        task.setStarted(0);
    }
}

File: app/src/main/kotlin/gargoyle/ct/task/impl/CTTimer.kt
package gargoyle.ct.task.impl

import gargoyle.ct.config.CTConfig
import gargoyle.ct.task.CTTaskUpdatable
import gargoyle.ct.task.helper.CTTimeHelper
import gargoyle.ct.util.util.className
import java.util.*

class CTTimer(timeHelper: CTTimeHelper, updatables: Iterable<CTTaskUpdatable>) {
    private val timer: Timer = Timer(CTTimer::class.className, true)
    private val timerTask: CTTimerTask

    constructor(timeHelper: CTTimeHelper, vararg updatables: CTTaskUpdatable) :
            this(timeHelper, listOf<CTTaskUpdatable>(*updatables))

    init {
        timerTask = CTTimerTask(timeHelper, updatables)
        timer.scheduleAtFixedRate(timerTask, CHECK_DELAY.toLong(), CHECK_PERIOD.toLong())
    }

    fun arm(config: CTConfig, currentMillis: Long): Unit =
        with(timerTask.task) {
            this.config = config
            this.started = currentMillis
        }

    fun terminate(): Unit = timer.cancel()

    fun unarm() {
        timerTask.task.started = 0
    }

    companion object {
        private const val CHECK_DELAY = 100
        private const val CHECK_PERIOD = 500
    }
}
